### PR TITLE
#2018140 deploy local changes

### PR DESCRIPTION
**Pull request by Fabustus**

### CODE REVIEW

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    /**
+     * Checks if the "deploy"-method is called and do nothing
+     */
+    @Test
+    void shouldCallDeploy()
+    {
+      List<String> uncommittedFiles = new ArrayList<>();
+      uncommittedFiles.add("test");
```
Comment (L1nc0ln) [inline nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:71]: You can use List.of("test") here instead of the two lines, it's more concise and mostly does the same (only difference is that the List.of list is immutable, but you don't need mutability in you test case)

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+      DeployLocalChangesAction deploySpy = spy(DeployLocalChangesAction.class);
+
+      doNothing().when(deploySpy).deploy(any());
+      doReturn(uncommittedFiles).when(deploySpy).getSourcesToDeploy(new HashSet<>());
+
+      deploySpy.actionPerformed(mock(ActionEvent.class));
+
+      verify(deploySpy).deploy(uncommittedFiles);
```
Comment (L1nc0ln) [inline nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:80]: You could add another test here that checks that deploy is not called if getSourcesToDeploy returns an empty list

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+     * Checks if no error is thrown when method is called
+     */
+    @Test
+    void shouldNotThrowIfActionPerformedIsCalled()
+    {
+      assertDoesNotThrow(() -> deployLocalChangesAction.actionPerformed(mock(ActionEvent.class)));
+    }
+
+    /**
+     * Checks if the "deploy"-method is called and do nothing
```
Comment (L1nc0ln) [inline nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:65]: The second part of the sentence does not really make sense here or at least does not add anything of value. Wouldn't "Checks if the "deploy"-method is called if there is at least one element to deploy" more fitting/accurate?

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  @Nested
+  class GetSourcesToDeploy
+  {
+    /**
+     * checks if an empty list is returned when parameter is empty
+     */
+    @Test
+    void shouldReturnEmptyList()
+    {
+      assertEquals(new ArrayList<>(), deployLocalChangesAction.getSourcesToDeploy(new HashSet<>()));
```
Comment (L1nc0ln) [inline nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:96]: You can use List.of() instead of new ArrayList here, same goes for the HashSet, Set.of exists

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  /**
+   * Checks the method {@link DeployLocalChangesAction#getSourcesToDeploy}
+   */
+  @Nested
+  class GetSourcesToDeploy
+  {
+    /**
+     * checks if an empty list is returned when parameter is empty
```
Comment (L1nc0ln) [inline nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:91]: when the input set is empty

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    void shouldReturnEmptyList()
+    {
+      assertEquals(new ArrayList<>(), deployLocalChangesAction.getSourcesToDeploy(new HashSet<>()));
+    }
+
+    /**
+     * checks if the empty provided list will not trigger the matcher-function
+     */
+    @Test
+    void shouldReturnEmptyListIfUncommitedChangesAreEmpty()
```
Comment (L1nc0ln) [inline nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:103]: I'd argue this test is superfluous since you are checking implementation behavior - if you change the way that you match the elements in the list at any point in the future, you will have to rewrite this test. It is preferably, to test if the expected output is returned, since you do not have to rewrite the test in case you change the implementation then

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+      }
+    }
+
+    /**
+     * checks if the function with the Regex returns the expected list
+     */
+    @Test
+    void shouldFindMatcherAndAddToList()
+    {
+      List<String> expected = new ArrayList<>();
```
Comment (L1nc0ln) [inline nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:122]: I'd use List.of and Set.of here as well. You can also try to test more than one input, so one test for no elements in the set, one for one element and one for multiple elements. Also maybe try some sensible Strings and some that may be troublesome (empty string, a string that does not match the expected input, like just "entity" here or something like that)

File: nbm/pom.xml [hunk 8]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0"
-    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (L1nc0ln) [inline nbm/pom.xml:2]: why did you change the ordering here? This shouldn't be wrong, but try to avoid any unnecessary changes in commits, since those obscure the actual changes

File: nbm/pom.xml [hunk 9]
```diff
@@ -25,6 +25,10 @@
       <groupId>de.adito.nb-modules</groupId>
       <artifactId>nb-adito-interface</artifactId>
     </dependency>
+    <dependency>
```
Comment (L1nc0ln) [inline nbm/pom.xml:28]: Is this dependency used/necessary?

File: pom.xml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0"
-    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (L1nc0ln) [inline pom.xml:2]: Same for unnecessary changes here

**Review by L1nc0ln — CHANGES_REQUESTED** [review]

File: nbm/src/main/java/de/adito/git/nbm/actions/DeployLocalChangesAction.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        .map(this::getSourcesToDeploy);
+
+    List<String> uncommittedFiles = uncommittedFilesObs.blockingFirst();
+    if (!uncommittedFiles.isEmpty())
+    {
+      deploy(uncommittedFiles);
+    }
+    else
+    {
+      INotificationFacade.INSTANCE.notify("Deploy", "There is nothing to deploy", true, null);
```
[the remark this answers is not in the document]
Comment (Fabustus) [reply nbm/src/main/java/de/adito/git/nbm/actions/DeployLocalChangesAction.java:66]: Not sure if this is what you want to do here. I want a baloon dialog in ADITO to show that there was nothing to deploy, if no local changes are made. Does NbBundle do the same?

File: nbm/pom.xml (hunk 8 above)
Comment (Fabustus) [reply nbm/pom.xml:2]: whoops, it did it automatically ^^

File: nbm/pom.xml (hunk 9 above)
Comment (Fabustus) [reply nbm/pom.xml:28]: Yeah for INotificationFacade

File: nbm/src/main/java/de/adito/git/nbm/actions/DeployLocalChangesAction.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (L1nc0ln) [reply nbm/src/main/java/de/adito/git/nbm/actions/DeployLocalChangesAction.java:66]: No, NbBundle just retrieves a string from a Bundle File, if there are Bundles for several languages and a Locale is given, the proper translation is used. You can use the NbBundle to retrieve the "Deploy" and "There is nothing to deploy" strings

**Review by L1nc0ln — CHANGES_REQUESTED** [review]

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java (hunk 6 above)
Comment (L1nc0ln) [reply nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:233]: If the parameterized test covers all cases of the tests it replaces then yes, it would replace them all. If it only covers some cases, keep the uncovered cases as separate tests

File: nbm/src/main/java/de/adito/git/nbm/actions/DeployLocalChangesAction.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    }
+  }
+
+  @VisibleForTesting
+  boolean isEnabled(@Nullable IRepository pRepository)
+  {
+    if (pRepository == null)
+      return false;
+
+    Observable<Object> res = pRepository
```
Comment (L1nc0ln) [inline nbm/src/main/java/de/adito/git/nbm/actions/DeployLocalChangesAction.java:66]: You should be able to make this an Observable<Boolean>, but you can also combine this with the return statement and directly return the result

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+   * }
+   */
+
+  /**
+   * Checks if the getIsEnabledObservable method returns the right enabled state
+   */
+  @Test
+  void shouldCallIsEnabled()
+  {
+    DeployLocalChangesAction deploySpy = spy(DeployLocalChangesAction.class);
```
Comment (L1nc0ln) [inline nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:96]: This test is basically only testing the RXJava Framework, not your code itself. I'd remove this test

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+    /**
+     * Parameterized test to check that the action is only available if one project is selected
+     *
+     * @param pRepository    the repository that will be checked
+     * @param expectedResult the expected result
+     */
+    @ParameterizedTest
+    @MethodSource("provideNodeArrays")
+    void shouldReturnExpectedResult(IRepository pRepository, boolean expectedResult)
```
Comment (L1nc0ln) [inline nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:154]: You could add a @NonNull annotation to the repository

**Review by L1nc0ln — CHANGES_REQUESTED** [review]

File: nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java (hunk 13 above)
Comment (L1nc0ln) [reply nbm/src/test/java/de/adito/git/nbm/actions/DeployLocalChangesActionTest.java:62]: The functionality of the getIsEnabledObservable method only consists of you calling RXJava, and you checking if the isEnabled method is called is implementation-specifc, meaning that if you rework the implementation of the method, you also have to rewrite the test
